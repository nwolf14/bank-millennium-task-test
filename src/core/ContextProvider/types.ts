// others
import { Theme } from './enums';

export type TContext = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
};
