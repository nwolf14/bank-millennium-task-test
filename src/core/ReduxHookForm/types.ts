// others
import { TField } from '../../store/reduxHookForm/types';
import { TObject } from './../../types';

export type TAsyncValidator = (
  value: boolean | number | string | Array<string>,
  subscribedFields: TObject<TField> | undefined
) => Promise<string>;

export type TSyncValidator = (
  value: boolean | number | string | Array<string>,
  subscribedFields: TObject<TField> | undefined
) => string;
