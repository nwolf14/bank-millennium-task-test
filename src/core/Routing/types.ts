import {
  ComponentClass,
  ComponentType,
  FunctionComponent,
  LazyExoticComponent,
  ReactElement,
} from 'react';
import { RouteProps } from 'react-router';

// others
import { RouteName } from './constants/routes';
import { TMainState } from '../../types';

export type TProtectedRouteProps = {
  guards?: Array<TGuard>;
  name: RouteName;
} & RouteProps;

export type TGuard = {
  guardCheck: (state: TMainState, routeProps?: RouteProps) => boolean;
  renderFallback: (
    routeProps: TProtectedRouteProps
  ) => JSX.Element | ReactElement;
};

export type TComponent =
  | ComponentType<{}>
  | LazyExoticComponent<FunctionComponent>
  | LazyExoticComponent<ComponentClass<{}, unknown>>
  | undefined;

export type TAppRouteData = {
  name: RouteName;
  path?: string;
  Component?: TComponent;
  guards?: Array<TGuard>;
};

export type TAppRoutesData = Array<TAppRouteData | TAppRouteData['name']>;
