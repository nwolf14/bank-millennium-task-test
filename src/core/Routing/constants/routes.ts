export enum RouteName {
  Dashboard = 'dashboard',
}

export const ROUTES = {
  [RouteName.Dashboard]: '/',
};
