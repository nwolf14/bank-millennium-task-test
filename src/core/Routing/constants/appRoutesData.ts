import { lazy } from 'react';

// components
const DashboardPage = lazy(
  () => import('../../../pages/DashboardPage/DashboardPage')
);

// core
import { RouteName } from './routes';

// others
import { TAppRoutesData } from '../types';

export const APP_ROUTES_DATA: TAppRoutesData = [
  {
    Component: DashboardPage,
    name: RouteName.Dashboard,
  }
];
