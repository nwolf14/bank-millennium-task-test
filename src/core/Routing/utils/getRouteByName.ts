// others
import { RouteName, ROUTES } from '../constants/routes';

export const getRouteByName = (name: RouteName): string => ROUTES[name];
