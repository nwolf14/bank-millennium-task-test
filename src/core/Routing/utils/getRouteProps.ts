// others
import { ROUTES } from '../constants/routes';
import { TAppRouteData, TProtectedRouteProps } from '../types';

const getRoutePath = ({ name }: TAppRouteData): TAppRouteData['path'] =>
  ROUTES[name];

export const getRouteProps = (
  routeData: TAppRouteData
): TProtectedRouteProps => ({
  path: getRoutePath(routeData),
  name: routeData.name,
  exact: true,
  component: routeData.Component,
  guards: routeData.guards || [],
});
