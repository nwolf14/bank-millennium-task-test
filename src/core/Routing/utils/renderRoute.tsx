// components
import ProtectedRoute from '../components/ProtectedRoute/ProtectedRoute';

// others
import { TAppRouteData } from '../types';

// utils
import { getRouteProps } from './getRouteProps';

export const renderRoute = (
  routeDataArg: TAppRouteData | TAppRouteData['name']
): JSX.Element => {
  const routeData: TAppRouteData =
    typeof routeDataArg === 'string' ? { name: routeDataArg } : routeDataArg;
  const routeProps = getRouteProps(routeData);

  return <ProtectedRoute {...routeProps} key={routeData.name} />;
};
