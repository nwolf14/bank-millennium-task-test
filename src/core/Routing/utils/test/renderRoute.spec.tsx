import configureStore from 'redux-mock-store';
import { ConnectedRouter } from 'connected-react-router';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import DashboardPage from '../../../../pages/DashboardPage/DashboardPage';

// core
import { renderRoute } from '../renderRoute';

// others
import { RouteName } from '../../constants/routes';
import { TAppRouteData } from '../../types';

// utils
import { history } from '../../../../utils/history';

const stateMock = {
  router: {
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
    },
    action: 'POP',
  },
};

const routeData: TAppRouteData = {
  Component: DashboardPage,
  name: RouteName.Dashboard,
};

describe('renderRoute', () => {
  const mockStore = configureStore();

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore(stateMock)}>
        <ConnectedRouter history={history}>
          {renderRoute(routeData)}
        </ConnectedRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
