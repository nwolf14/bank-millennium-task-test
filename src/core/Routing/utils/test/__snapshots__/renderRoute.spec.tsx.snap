// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderRoute matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <a
      class=""
      href="/strategy-visual-builder"
    >
      Strategy Visual Builder
    </a>
  </div>
</DocumentFragment>
`;
